modules/javafx.graphics/src/main/java/com/sun/javafx/css/media/MediaRule.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.css.media;
+
+import javafx.css.StyleConverter;
+import javafx.scene.Scene;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * CSS media rules are @-rules that contain media queries. A media query tests the "external"
+ * configuration of a {@link Scene}, and is independent of its scene graph content.
+ */
+public final class MediaRule {
+
+    private final List<MediaQuery> queries;
+    private final MediaRule parent;
+    private final int hash;
+
+    public MediaRule(List<MediaQuery> queries, MediaRule parent) {
+        this.queries = List.copyOf(queries);
+        this.parent = parent;
+        this.hash = queries.hashCode();
+    }
+
+    /**
+     * Returns the list of media queries.
+     *
+     * @return the list of media queries
+     */
+    public List<MediaQuery> getQueries() {
+        return queries;
+    }
+
+    /**
+     * Returns the parent media rule if this media rule is nested.
+     *
+     * @return the parent {@code MediaRule}, or {@code null}
+     */
+    public MediaRule getParent() {
+        return parent;
+    }
+
+    /**
+     * Evaluates whether any of the media queries is {@code true}.
+     *
+     * @param context the evaluation context
+     * @return {@code true} if any of the media queries is {@code true} or if the list is empty,
+     *         {@code false} otherwise
+     */
+    public boolean evaluate(MediaQueryContext context) {
+        if (parent != null && !parent.evaluate(context)) {
+            return false;
+        }
+
+        if (queries.isEmpty()) {
+            return true;
+        }
+
+        for (int i = 0, max = queries.size(); i < max; i++) {
+            if (queries.get(i).evaluate(context)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public void writeBinary(DataOutputStream stream, StyleConverter.StringStore stringStore) throws IOException {
+        stream.writeInt(queries.size());
+
+        for (MediaQuery query : queries) {
+            MediaQuerySerializer.writeBinary(query, stream, stringStore);
+        }
+
+        stream.writeBoolean(parent != null);
+
+        if (parent != null) {
+            parent.writeBinary(stream, stringStore);
+        }
+    }
+
+    public static MediaRule readBinary(DataInputStream stream, String[] strings) throws IOException {
+        int size = stream.readInt();
+        var queries = new MediaQuery[size];
+
+        for (int i = 0; i < size; i++) {
+            queries[i] = MediaQuerySerializer.readBinary(stream, strings);
+        }
+
+        boolean hasParent = stream.readBoolean();
+
+        return new MediaRule(List.of(queries), hasParent ? readBinary(stream, strings) : null);
```

Comment:
the exceptions in java have improved, but there is 3 statements in this line.
anyone who ever debugged code after the fact using two-week old logs would say: please keep one statement per line, if possible.